src/routes/settings/OpenAI.svelte

```diff
@@ -0,0 +1,87 @@
+<script lang="ts">
+	import { onMount } from 'svelte';
+
+	import LL from '$i18n/i18n-svelte';
+	import { OpenAIStrategy } from '$lib/chat/openai';
+	import Badge from '$lib/components/Badge.svelte';
+	import Button from '$lib/components/Button.svelte';
+	import FieldHelp from '$lib/components/FieldHelp.svelte';
+	import FieldInput from '$lib/components/FieldInput.svelte';
+	import Fieldset from '$lib/components/Fieldset.svelte';
+	import P from '$lib/components/P.svelte';
+	import { settingsStore } from '$lib/localStorage';
+
+	let openaiURL: URL | null = null;
+
+	const DETAULT_OPENAI_SERVER = 'http://localhost:11434/v1/';
+
+	let openai = new OpenAIStrategy();
+	let openaiServer = $settingsStore.openaiServer || DETAULT_OPENAI_SERVER;
+	let openaiApiKey = $settingsStore.openaiApiKey || '';
+	let openaiServerStatus: 'connected' | 'disconnected' = 'disconnected';
+
+	$: settingsStore.update((settings) => ({ ...settings, openaiServer, openaiApiKey }));
+
+	async function getModelsList(): Promise<void> {
+		try {
+			await openai.getModels();
+			openaiServerStatus = 'connected';
+		} catch {
+			openaiServerStatus = 'disconnected';
+		}
+	}
+
+	onMount(async () => {
+		// Get the current URL and set the default server
+		openaiURL = new URL(window.location.href);
+		if (openaiURL.port) {
+			openaiURL = new URL(
+				`${openaiURL.protocol}//${openaiURL.hostname}${openaiURL.pathname}${openaiURL.search}${openaiURL.hash}`
+			);
+		}
+
+		await getModelsList();
+	});
+</script>
+
+<Fieldset>
+	<P><strong>OpenAI</strong></P>
+	<FieldInput
+		name="baseUrl"
+		label={$LL.baseUrl()}
+		placeholder={DETAULT_OPENAI_SERVER}
+		bind:value={openaiServer}
+		on:keyup={getModelsList}
+	>
```

Comment:
First, in `onMount`:
- let's remove the `getModelsList` call when the component mounts.
- `openAiURL` should be `https://api.openai.com/v1`, not `localhost`.

Then, one thing I noticed is that after successfully connecting to OpenAI, if I leave away from **Settings** and then come back, it's always showing as `disconnected` first.

I understand that this is how the Ollama implementation was working but now that the connection to OpenAI is much slower it's very noticeable and feels like mistake.

I'm wondering if we should even remove the `onKeyUp` events and add a button to intentionally "Connect" to the server `onClick`, similar to the one for pulling Ollama models.

If we add a button then we can probably remove the `connected/disconnected` badges and simply show a `Connecting`, `Connected/Disconnected` toast alerts (or maybe even something more useful like status updates, for example: `Retrieving models`).

Anyways, let's not add a button yet, I'm just thinking out loud.

Let's fix all of the other stuff first and we'll come back to this later once we have a better feel for the rest of the UX.

Reply:
Thanks for the feedback! I've improved several things about the OpenAI config, please let me know what you think